### PR TITLE
Torch Model and tests for smiles2vec

### PR DESCRIPTION
**Pull request by harishwar017**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/smiles2vec.py [hunk 1]
```diff
@@ -177,3 +182,103 @@ def forward(self, smiles_seqs: List):
         else:
             Logits = Logits.view(-1, self.n_tasks, 1)
             return Logits
+
+
+class Smiles2VecModel(TorchModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/smiles2vec.py:187]: Docstrings and usage example.

File: deepchem/models/torch_models/smiles2vec.py [hunk 2]
```diff
@@ -177,3 +182,103 @@ def forward(self, smiles_seqs: List):
         else:
             Logits = Logits.view(-1, self.n_tasks, 1)
             return Logits
+
+
```
Comment (rbharath) [inline deepchem/models/torch_models/smiles2vec.py:186]: Please add this model to the .rst docs. Also add it to the model cheatsheet csv file

File: deepchem/models/torch_models/smiles2vec.py (hunk 1 above)
Comment (rbharath) [inline deepchem/models/torch_models/smiles2vec.py:187]: Please doublecheck it is exactly the same name as TF. Also make sure all the arguments overlap

**Review by rbharath — COMMENTED** [review]
Have docs added. This is close to ready (code/tests looks good)

File: deepchem/models/torch_models/smiles2vec.py [hunk 3]
```diff
@@ -177,3 +182,103 @@ def forward(self, smiles_seqs: List):
...
+        else:
+            loss = L2Loss()
+            output_types = ['prediction']
+
+        super(Smiles2VecModel, self).__init__(self.model,
+                                              loss=loss,
+                                              output_types=output_types,
+                                              **kwargs)
+
+    def default_generator(self,
```
Comment (rbharath) [inline deepchem/models/torch_models/smiles2vec.py:261]: Why is this being overwritten?

Comment (rbharath) [reply deepchem/models/torch_models/smiles2vec.py:261]: Check the other torch models

File: deepchem/models/torch_models/tests/test_smiles2vec.py [hunk 4]
```diff
@@ -77,9 +83,98 @@ def test_smiles2vec_model():
...
+                       max_seq_len=max_seq_len,
+                       n_tasks=n_tasks)
+
     input = torch.randint(low=0, high=len(char_to_idx), size=(1, max_seq_len))
     # Ex: input = torch.tensor([[32,32,32,32,32,32,25,29,15,17,29,29,32,32,32,32,32,32,32,32]])
 
     logits = model.forward(input)
     assert np.shape(logits) == (1, n_tasks, 1)
+
+
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_smiles2vec.py:96]: Check if input/output for TF and Torch match with same weights

Comment (harishwar017) [reply deepchem/models/torch_models/tests/test_smiles2vec.py:98]: yes I have added the test

File: deepchem/models/torch_models/smiles2vec.py (hunk 3 above)
Comment (harishwar017) [reply deepchem/models/torch_models/smiles2vec.py]: removed overwritten default generator

File: deepchem/models/torch_models/smiles2vec.py (hunk 2 above)
Comment (harishwar017) [reply deepchem/models/torch_models/smiles2vec.py:189]: done adding

File: deepchem/models/torch_models/smiles2vec.py (hunk 1 above)
Comment (harishwar017) [reply deepchem/models/torch_models/smiles2vec.py:190]: just added them in the last commit

Comment (harishwar017) [reply deepchem/models/torch_models/smiles2vec.py:190]: checked them, and they do overlap

**Review by Shiva-sankaran — CHANGES_REQUESTED** [review]
LGTM, after minor corrections.

File: deepchem/models/torch_models/tests/test_smiles2vec.py [hunk 5]
```diff
@@ -77,9 +85,284 @@ def test_smiles2vec_model():
...
+    ]
+
+    def regroup_params_gru(weight_or_bias_gru, axis=0):
+        assert len(weight_or_bias_gru.shape
+                  ) == 2 and weight_or_bias_gru.shape[axis] % 3 == 0
+        # change params for the 3 gates from tf order(z,r,h) into torch order(r,z,h)
+        [z, r, h] = np.split(weight_or_bias_gru, 3, axis=axis)
+        return np.concatenate((r, z, h), axis=axis)
+
+    torch_model = Smiles2VecModel(char_to_idx=char_to_idx,
```
[the remark this answers is not in the document]
Comment (harishwar017) [reply deepchem/models/torch_models/tests/test_smiles2vec.py]: removed it

File: deepchem/models/torch_models/smiles2vec.py [hunk 6]
```diff
@@ -177,3 +185,156 @@ def forward(self, smiles_seqs: List):
...
+        rnn_sizes: list[int], default [224, 384]
+            Number of hidden units in the RNN cells
+        mode: str, default regression
+            Whether to use model for regression or classification
+        device: torch.device, optional (default None)
+            the device on which to run computations.  If None, a device is
+            chosen automatically
+        """
+        self.n_tasks = n_tasks
+        if device is None:
```
[the remark this answers is not in the document]
Comment (harishwar017) [reply deepchem/models/torch_models/smiles2vec.py]: got it, removed that from Smiles2VecModel and sent it through **kwargs

**Review by rbharath — COMMENTED** [review]
LGTM once CI passes (may need to be rebased after torch patch)

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**